expose `line` function

raidancampbell opened the pull request:
already exists in framebuf, can be used today by reaching through the oled object with `oled.framebuf.line(16,16,32,32, 1)`

raidancampbell committed:
expose `line` function

already exists in framebuf, can be used today by reaching through the oled object with `oled.framebuf.line(16,16,32,32, 1)`

changed variable names to conform to conventions

Removed trailing whitespace, and attempted to disable `too-many-arguments` linting

tannewt commented on adafruit_ssd1306.py:
```diff
@@ -130,6 +130,12 @@ def text(self, string, xpos, ypos, col=1):
         """Place text on display"""
         self.framebuf.text(string, xpos, ypos, col)
 
+    #pylint: disable-msg=too-many-arguments
```
I think you only need this `disable` once if you put it in the method after the comment.

raidancampbell replied:
whoops I meant to re-enable it after this block, I saw some other hunk of code that did that.  Copy/pasted a bit too much, fixing...

raidancampbell replied:
resolved via 4bc78f8a7ffac812b523f3d9d7cba2f8215d658b

tannewt reviewed (changes requested):
One thing to reduce lint clutter.

raidancampbell committed:
Update adafruit_ssd1306.py

minor fix for linting hints. re-enable `too-many-arguments` after the block in question

deshipu commented on adafruit_ssd1306.py:
```diff
@@ -130,6 +130,12 @@ def text(self, string, xpos, ypos, col=1):
+        """Draw a line from initial to final point"""
+        self.framebuf.line(xpos0, ypos0, xpos1, ypos1, col)
+    #pylint: enable-msg=too-many-arguments
+
```
I think it would be much more efficient to just add `self.line = self.framebuf.line` to `__init__` — and same for all of the other functions.

deshipu commented:
I'm not sure we want to have to do this in every single driver, and then update those drivers when the `framebuf` gets new methods or changes their signatures, etc. — I think we should simply document how to call the methods on the framebuffer itself.

raidancampbell commented:
Never thought about that: I agree.  I'm unsure how to fully test this however: the lines between `py` and `mpy` are pretty blurred for me.  It seems like this code automatically interfaces with both `mpy` (compiled) and `py` (interpreted) code, depending on which is available.

I have an [open PR at the micropython-adafruit-framebuf repo](https://github.com/adafruit/micropython-adafruit-framebuf/pull/3) for mirroring the C logic in python.  Ideally I'd like that to be accepted before this gets accepted: I've tested that code as working fine.

I haven't ever been able to use the python-backed framebuf: no matter what I try, the actual import is the `mpy` framebuf.  If someone can give me a hand in making that work, I'm happy to unify the `mpy`/`py` framebuffers and clean up this library's interface.

tannewt commented:
@raidancampbell I think the tricky thing here is that the built in framebuf is actually [implemented in C](https://github.com/adafruit/circuitpython/blob/master/extmod/modframebuf.c#L575) and there is a second Python implemented version that can also be converted to bytecode (known as mpy). To test, I'd recommend renaming the python version so that it doesn't conflict and then changing the import in the library.

It'd be awesome to get the framebuf module reorganized in CircuitPython so that its in `shared-bindings` like [`struct`](https://github.com/adafruit/circuitpython/tree/master/shared-bindings/struct).

tannewt commented:
@raidancampbell Are you hoping to get back to this and test it?

raidancampbell commented:
I implemented and tested, but hadn't gotten around to replying. The "cleaner" method of just mapping the functions is a breaking change: the existing wrapper functions specify a default color if none is given. The wrapped function requires the color to be given.  If everyone is okay with that, I'll update with the function mapping to replace wrapping.

raidancampbell committed:
Implement function mapping and specify the updated framebuffer to be used

raidancampbell committed:
Merge remote-tracking branch 'origin/patch-1' into patch-1

# Conflicts:
#	adafruit_ssd1306.py

raidancampbell committed:
Merge resolution
- I might have tripped myself up in git here...

tannewt commented on adafruit_ssd1306.py:
```diff
@@ -35,6 +35,19 @@ class _SSD1306:
         self.framebuf = framebuffer
+        # note this is a breaking change:
+        # previously these functions were wrapped
+        # with a default color of "1"
```
This comment should go in the PR message not the code. I bet col=1 below is actually column not color too.

tannewt reviewed (changes requested):
One minor thing.

tannewt commented:
I'm ok with the breaking change. It seems very minor to me. Ping me on [Discord](https://adafru.it/discord) is you need git help.

raidancampbell commented:
closing to reopen in #9
